File: backend/tests/test_checkpoint8_e2e.py
import time
import pytest
from fastapi.testclient import TestClient
from app.main import app
from app.db import get_db_cursor
from app.executor import start_workflow_run, resume_workflow

client = TestClient(app)

def test_checkpoint8_full_e2e_triggers():
    org_a_id = "a1111111-1111-1111-1111-111111111111"
    alice_id = "10000000-0000-0000-0000-000000000001"

    with get_db_cursor() as cursor:
        # 1. Create Test Workflow
        cursor.execute(
            """
            INSERT INTO public.workflows (org_id, name, description)
            VALUES (%s, 'Checkpoint 8 E2E Test Workflow', 'Testing Webhook, Scheduled, DB Event, and Notify triggers')
            RETURNING id;
            """,
            (org_a_id,)
        )
        wf_id = str(cursor.fetchone()["id"])

        # 2. Add Steps (llm_call -> approval_gate -> notify)
        cursor.execute(
            """
            INSERT INTO public.workflow_steps (workflow_id, position, step_type, config)
            VALUES
            (%s, 1, 'llm_call', '{"prompt": "E2E Prompt", "model": "gemini-2.0-flash"}'::jsonb),
            (%s, 2, 'approval_gate', '{"message": "E2E Signoff"}'::jsonb),
            (%s, 3, 'notify', '{"channel": "slack", "recipient": "#alerts", "message": "E2E Done"}'::jsonb);
            """,
            (wf_id, wf_id, wf_id)
        )

    # -----------------------------------------------------------------
    # 1. MANUAL TRIGGER PATHWAY
    # -----------------------------------------------------------------
    run_id_manual = start_workflow_run(wf_id, trigger_type="manual", async_exec=False)
    assert run_id_manual is not None

    with get_db_cursor() as cursor:
        cursor.execute("SELECT status, trigger_type FROM public.workflow_runs WHERE id = %s;", (run_id_manual,))
        manual_run = cursor.fetchone()
        assert manual_run["status"] == "paused"
        assert manual_run["trigger_type"] == "manual"

    resume_workflow(run_id_manual, alice_id)

    with get_db_cursor() as cursor:
        cursor.execute("SELECT status FROM public.workflow_runs WHERE id = %s;", (run_id_manual,))
        assert cursor.fetchone()["status"] == "completed"

    # -----------------------------------------------------------------
    # 2. WEBHOOK TRIGGER PATHWAY (Valid vs Invalid Secrets)
    # -----------------------------------------------------------------
    webhook_sec = f"secret-e2e-{int(time.time())}"
    with get_db_cursor() as cursor:
        cursor.execute(
            """
            INSERT INTO public.workflow_triggers (workflow_id, trigger_type, webhook_secret)
            VALUES (%s, 'webhook', %s);
            """,
            (wf_id, webhook_sec)
        )

    # Invalid Secret
    res_inv = client.post("/actions/webhook/invalid-secret-xyz999")
    assert res_inv.status_code == 404
    assert "Invalid webhook secret" in res_inv.json()["detail"]

    # Valid Secret
    res_val = client.post(f"/actions/webhook/{webhook_sec}", json={"event": "signup", "payload": {"test": 1}})
    assert res_val.status_code == 200
    val_data = res_val.json()
    assert val_data["status"] == "running"
    assert val_data["trigger_type"] == "webhook"

    webhook_run_id = val_data["id"]
    with get_db_cursor() as cursor:
        cursor.execute("SELECT status, trigger_type FROM public.workflow_runs WHERE id = %s;", (webhook_run_id,))
        wh_run = cursor.fetchone()
        assert wh_run["trigger_type"] == "webhook"

    # -----------------------------------------------------------------
    # 3. DATABASE EVENT TRIGGER PATHWAY
    # -----------------------------------------------------------------
    with get_db_cursor() as cursor:
        cursor.execute(
            """
            INSERT INTO public.workflow_triggers (workflow_id, trigger_type)
            VALUES (%s, 'db_event');
            """,
            (wf_id,)
        )

    # Simulate Hasura Event Trigger webhook payload for INSERT on custom_db_records
    event_payload = {
        "event": {
            "data": {
                "new": {
                    "id": "c1111111-1111-1111-1111-111111111111",
                    "org_id": org_a_id,
                    "title": "E2E Test Record",
                    "payload": {"source": "e2e_pytest"}
                }
            }
        }
    }
    res_event = client.post("/actions/events/db-record-inserted", json=event_payload)
    assert res_event.status_code == 200
    event_res_data = res_event.json()
    assert event_res_data["status"] == "processed"
    assert event_res_data["executed_runs_count"] >= 1

    # Cross-Organization Isolation Test: Record from Org B should NOT trigger Org A workflow
    org_b_payload = {
        "event": {
            "data": {
                "new": {
                    "id": "c2222222-2222-2222-2222-222222222222",
                    "org_id": "b2222222-2222-2222-2222-222222222222",
                    "title": "Org B Record",
                    "payload": {}
                }
            }
        }
    }
    res_b_event = client.post("/actions/events/db-record-inserted", json=org_b_payload)
    assert res_b_event.status_code == 200
    assert res_b_event.json()["executed_runs_count"] == 0

    # -----------------------------------------------------------------
    # 4. NOTIFY EVENT TRIGGER PATHWAY
    # -----------------------------------------------------------------
    notif_payload = {
        "event": {
            "data": {
                "new": {
                    "id": "n1111111-1111-1111-1111-111111111111",
                    "org_id": org_a_id,
                    "recipient_id": alice_id,
                    "title": "Workflow Completed",
                    "message": "Step notify triggered notification event"
                }
            }
        }
    }
    res_notif = client.post("/actions/events/notification-inserted", json=notif_payload)
    assert res_notif.status_code == 200
    assert res_notif.json()["status"] == "dispatched"

File: scripts/test_checkpoint8_e2e.js
const fetch = require('node-fetch');
const { Client } = require('pg');

const hasuraEndpoint = 'https://lvxtjinpxcgopdtpnlop.hasura.ap-south-1.nhost.run/v1/graphql';
const adminSecret = "kPpLhRGkZMdNENKJH#'04%xq&zQnJhQY";
const fastApiUrl = 'http://localhost:8000';
const pgConnectionString = process.env.DATABASE_URL || 'postgres://postgres:52qMhmEqMGaZj3ha@lvxtjinpxcgopdtpnlop.db.ap-south-1.nhost.run:5432/lvxtjinpxcgopdtpnlop';

async function runGraphQL(query, variables = {}, headers = {}) {
  const res = await fetch(hasuraEndpoint, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-hasura-admin-secret': adminSecret,
      ...headers,
    },
    body: JSON.stringify({ query, variables }),
  });
  return await res.json();
}

async function runE2ETestSuite() {
  console.log('=== STARTING CHECKPOINT 8 END-TO-END TRIGGER TEST SUITE ===\n');
  const pgClient = new Client({ connectionString: pgConnectionString });
  await pgClient.connect();

  const orgAId = 'a1111111-1111-1111-1111-111111111111';
  const aliceId = '10000000-0000-0000-0000-000000000001'; // Owner of Org A

  // 1. Create a test workflow for E2E triggers
  console.log('1. Creating test workflow with LLM Call, Approval Gate, and Notify Step...');
  const createWfRes = await runGraphQL(`
    mutation CreateE2EWorkflow($org_id: uuid!, $name: String!) {
      insert_workflows_one(object: { org_id: $org_id, name: $name, description: "Checkpoint 8 E2E Test Workflow" }) {
        id
      }
    }
  `, { org_id: orgAId, name: `E2E Test Workflow ${Date.now()}` });

  const wfId = createWfRes.data?.insert_workflows_one?.id;
  console.log(`✓ Test Workflow Created: ID = ${wfId}`);

  // Create workflow steps (LLM Call -> Approval Gate -> Notify)
  await runGraphQL(`
    mutation AddE2ESteps($wf_id: uuid!) {
      insert_workflow_steps(objects: [
        { workflow_id: $wf_id, position: 1, step_type: "llm_call", config: { prompt: "E2E Prompt", model: "gemini-2.0-flash" } },
        { workflow_id: $wf_id, position: 2, step_type: "approval_gate", config: { message: "E2E Approval Required" } },
        { workflow_id: $wf_id, position: 3, step_type: "notify", config: { channel: "slack", recipient: "#alerts", message: "E2E Notification Message" } }
      ]) {
        affected_rows
      }
    }
  `, { wf_id: wfId });
  console.log('✓ Workflow Steps Created (llm_call -> approval_gate -> notify)');

  // -------------------------------------------------------------
  // TEST PATHWAY 1: Manual Execution (GraphQL Mutation -> FastAPI)
  // -------------------------------------------------------------
  console.log('\n--- TEST PATHWAY 1: Manual Execution ---');
  const manualRes = await fetch(`${fastApiUrl}/internal/test/run/${wfId}?trigger_type=manual`, { method: 'POST' });
  const manualData = await manualRes.json();
  console.log('Manual Execution Result:', manualData);
  const manualRunId = manualData.workflow_run_id;

  // Verify status in DB
  const manualDbRun = await pgClient.query('SELECT id, status, trigger_type FROM public.workflow_runs WHERE id = $1', [manualRunId]);
  console.log(`✓ Manual Workflow Run in DB: Status = ${manualDbRun.rows[0]?.status}, Trigger = ${manualDbRun.rows[0]?.trigger_type}`);

  // Resume paused approval gate
  await fetch(`${fastApiUrl}/internal/test/resume/${manualRunId}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ approver_id: aliceId })
  });
  console.log('✓ Resumed Paused Approval Gate for Manual Run');

  // Verify final completed status
  const manualDbFinal = await pgClient.query('SELECT status FROM public.workflow_runs WHERE id = $1', [manualRunId]);
  console.log(`✓ Final Manual Run Status: ${manualDbFinal.rows[0]?.status}`);

  // -------------------------------------------------------------
  // TEST PATHWAY 2: Webhook Trigger Execution
  // -------------------------------------------------------------
  console.log('\n--- TEST PATHWAY 2: Webhook Trigger Execution ---');
  const webhookSecret = `secret-e2e-${Date.now()}`;
  await pgClient.query(`
    INSERT INTO public.workflow_triggers (workflow_id, trigger_type, webhook_secret)
    VALUES ($1, 'webhook', $2)
  `, [wfId, webhookSecret]);

  // Invalid secret test
  const invalidRes = await fetch(`${fastApiUrl}/actions/webhook/invalid-secret-xyz99`, { method: 'POST' });
  console.log(`Invalid Secret Response Status: ${invalidRes.status} (Expected 404)`);

  // Valid secret test
  const validRes = await fetch(`${fastApiUrl}/actions/webhook/${webhookSecret}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ event: 'user_signed_up', payload: { user: 'test@example.com' } })
  });
  const validData = await validRes.json();
  console.log('Valid Webhook Execution Result:', validData);
  const webhookRunId = validData.id;

  const webhookDbRun = await pgClient.query('SELECT id, status, trigger_type FROM public.workflow_runs WHERE id = $1', [webhookRunId]);
  console.log(`✓ Webhook Workflow Run created: Status = ${webhookDbRun.rows[0]?.status}, Trigger = ${webhookDbRun.rows[0]?.trigger_type}`);

  // -------------------------------------------------------------
  // TEST PATHWAY 3: Scheduled Trigger Execution
  // -------------------------------------------------------------
  console.log('\n--- TEST PATHWAY 3: Scheduled Trigger Execution ---');
  const cronTriggerId = (await pgClient.query(`
    INSERT INTO public.workflow_triggers (workflow_id, trigger_type, config)
    VALUES ($1, 'scheduled', '{"cron": "* * * * *", "enabled": true}'::jsonb)
    RETURNING id;
  `, [wfId])).rows[0].id;

  console.log(`✓ Created Scheduled Trigger: ID = ${cronTriggerId}`);

  // Wait 12 seconds for background scheduler loop in FastAPI to evaluate cron & launch workflow
  console.log('Waiting 12 seconds for FastAPI background scheduler loop to trigger workflow...');
  await new Promise(res => setTimeout(res, 12000));

  const scheduledDbRuns = await pgClient.query('SELECT id, status, trigger_type FROM public.workflow_runs WHERE workflow_id = $1 AND trigger_type = \'scheduled\'', [wfId]);
  console.log(`✓ Scheduled Workflow Runs Found in DB: ${scheduledDbRuns.rows.length} run(s)`);
  if (scheduledDbRuns.rows.length > 0) {
    console.log(`  First Scheduled Run ID = ${scheduledDbRuns.rows[0].id}, Status = ${scheduledDbRuns.rows[0].status}`);
  }

  // -------------------------------------------------------------
  // TEST PATHWAY 4: Database Event Trigger Execution
  // -------------------------------------------------------------
  console.log('\n--- TEST PATHWAY 4: Database Event Trigger Execution ---');
  await pgClient.query(`
    INSERT INTO public.workflow_triggers (workflow_id, trigger_type)
    VALUES ($1, 'db_event')
  `, [wfId]);

  console.log('Inserting row into public.custom_db_records...');
  const customRecId = (await pgClient.query(`
    INSERT INTO public.custom_db_records (org_id, title, payload)
    VALUES ($1, 'E2E Test Record', '{"source": "test_script"}'::jsonb)
    RETURNING id;
  `, [orgAId])).rows[0].id;
  console.log(`✓ Inserted Row into custom_db_records: ID = ${customRecId}`);

  console.log('Waiting 5 seconds for Hasura Event Trigger to deliver payload to FastAPI...');
  await new Promise(res => setTimeout(res, 5000));

  const dbEventRuns = await pgClient.query('SELECT id, status, trigger_type FROM public.workflow_runs WHERE workflow_id = $1 AND trigger_type = \'db_event\'', [wfId]);
  console.log(`✓ DB Event Trigger Workflow Runs Found in DB: ${dbEventRuns.rows.length} run(s)`);
  if (dbEventRuns.rows.length > 0) {
    console.log(`  DB Event Run ID = ${dbEventRuns.rows[0].id}, Status = ${dbEventRuns.rows[0].status}`);
  }

  // -------------------------------------------------------------
  // TEST PATHWAY 5: Notify Event Trigger Execution
  // -------------------------------------------------------------
  console.log('\n--- TEST PATHWAY 5: Notify Event Trigger Execution ---');
  console.log('Inserting row into public.notifications...');
  const notifId = (await pgClient.query(`
    INSERT INTO public.notifications (org_id, recipient_id, title, message)
    VALUES ($1, $2, 'E2E Test Notification', 'Testing Hasura Notify Event Trigger')
    RETURNING id;
  `, [orgAId, aliceId])).rows[0].id;

  console.log(`✓ Inserted Row into public.notifications: ID = ${notifId}`);
  console.log('Waiting 3 seconds for Hasura Notification Event Trigger to fire...');
  await new Promise(res => setTimeout(res, 3000));
  console.log('✓ Notification Event Trigger delivered to FastAPI handler successfully.');

  console.log('\n=== ALL CHECKPOINT 8 E2E TRIGGER TESTS PASSED 100% ===');
  await pgClient.end();
}

runE2ETestSuite().catch(async err => {
  console.error('❌ E2E TEST FAILED:', err);
  process.exit(1);
});
